src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/EventProcessorService.cs

```diff
@@ -0,0 +1,488 @@
+            this.CheckpointManager = new ReliableDictionaryCheckpointMananger(this.ServiceStateManager);
+            this.EventHubClientFactory = new EventHubWrappers.EventHubClientFactory();
+            this.TestMode = false;
+
+            this.internalCanceller = new CancellationTokenSource();
+        }
+
+        /// <summary>
+        /// Set processing options in the constructor.
+        /// </summary>
+        public EventProcessorOptions Options { get; set; }
+
+        /// <summary>
+        /// Optionally provide a user implementation of the event processor factory.
+        /// </summary>
+        public IEventProcessorFactory EventProcessorFactory { get; set; }
+
+        /// <summary>
+        /// Optionally provide a user implementation of the checkpoint manager.
+        /// </summary>
+        public ICheckpointMananger CheckpointManager { get; set; }
+
+        /// <summary>
+        /// For testing purposes.
+        /// </summary>
+        public EventHubWrappers.IEventHubClientFactory EventHubClientFactory { get; set; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public bool TestMode { get; set; }
+
+        /// <summary>
+        /// Called by Service Fabric.
+        /// </summary>
+        /// <param name="fabricCancellationToken"></param>
+        /// <returns></returns>
+        public async Task RunAsync(CancellationToken fabricCancellationToken)
+        {
+            try
+            {
+                using (CancellationTokenSource linkedCanceller = CancellationTokenSource.CreateLinkedTokenSource(fabricCancellationToken, this.internalCanceller.Token))
+                {
+                    this.linkedCancellationToken = linkedCanceller.Token;
+                    await InnerRunAsync();
+                    this.Options.NotifyOnShutdown(null);
+                }
+            }
+            catch (Exception e)
+            {
+                // If InnerRunAsync throws, that is intended to be a fatal exception for this instance.
+                // Catch it here just long enough to log and notify, then rethrow.
+
+                EventProcessorEventSource.Current.Message("THROWING OUT: {0}", e);
+                this.Options.NotifyOnShutdown(e);
+                throw e;
+            }
+        }
+
+        private async Task InnerRunAsync()
+        {
+            EventHubWrappers.IEventHubClient ehclient = null;
+            EventHubWrappers.IPartitionReceiver receiver = null;
+
+            try
+            {
+                //
+                // General startup tasks.
+                //
+                await PartitionStartup(this.linkedCancellationToken);
+
+                //
+                // Create EventHubClient and check partition count.
+                //
+                Exception lastException = null;
+                EventProcessorEventSource.Current.Message("Creating event hub client");
+                for (int i = 0; i < Constants.RetryCount; i++)
```

Comment:
>for (int i = 0; i < Constants.RetryCount; i++) [](start = 16, length = 46)

Seems we have couple similar retry blocks. Better to move them to a try-catch delegate.

Reply:
Unified the pattern in a method that takes a delegate.